File: envoy/proto-java/src/main/java/io/hydrosphere/serving/service/AuthorityReplacerInterceptor.java
package io.hydrosphere.serving.service;

import io.grpc.*;

public class AuthorityReplacerInterceptor implements ClientInterceptor {

    public static final CallOptions.Key<String> DESTINATION_KEY = CallOptions.Key.of("destination", "localhost");

    @Override
    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method,
                                                               CallOptions callOptions, Channel next) {

        String destination = callOptions.getOption(DESTINATION_KEY);
        CallOptions newCallOptions = callOptions.withAuthority(destination);
        return next.newCall(method, newCallOptions);
    }
}

File: manager/src/main/scala/io/hydrosphere/serving/manager/grpc/manager/AuthorityReplacerInterceptor.scala
package io.hydrosphere.serving.manager.grpc.manager

import io.grpc._

class AuthorityReplacerInterceptor extends ClientInterceptor {

  override def interceptCall[ReqT, RespT](
    method: MethodDescriptor[ReqT, RespT],
    callOptions: CallOptions,
    next: Channel
  ): ClientCall[ReqT, RespT] = {
    val destination = callOptions.getOption(AuthorityReplacerInterceptor.DESTINATION_KEY)
    val newCallOptions = callOptions.withAuthority(destination)
    next.newCall(method, newCallOptions)
  }

}

object AuthorityReplacerInterceptor {
  val DESTINATION_KEY: CallOptions.Key[String] = CallOptions.Key.of("destination", "localhost")
}